src/vstest.console/InProcessVsTestConsoleWrapper.cs

```diff
@@ -0,0 +1,1101 @@
+            var upcastCustomLauncher = (ITestHostLauncher3?)customLauncher;
+
+            TestRequestManager?.RunTests(
+                testRunPayload,
+                upcastCustomLauncher,
+                new RunHandlerToEventsRegistrarAdapter(testRunEventsHandler),
+                new ProtocolConfig { Version = _highestSupportedVersion });
+        }
+        catch (Exception ex)
+        {
+            EqtTrace.Error("InProcessVsTestConsoleWrapper.RunTestsWithCustomTestHost: Exception occurred: " + ex);
+            var testRunCompleteArgs = new TestRunCompleteEventArgs(null, false, true, ex, null, null, TimeSpan.MinValue);
+
+            testRunEventsHandler.HandleLogMessage(TestMessageLevel.Error, ex.ToString());
+            testRunEventsHandler.HandleTestRunComplete(testRunCompleteArgs, null, null, null);
+        }
+
+        _testPlatformEventSource.TranslationLayerExecutionStop();
+    }
+
+    /// <inheritdoc/>
+    public void RunTestsWithCustomTestHost(
+        IEnumerable<TestCase> testCases,
+        string? runSettings,
+        ITestRunEventsHandler testRunEventsHandler,
+        ITestHostLauncher? customTestHostLauncher)
+    {
+        RunTestsWithCustomTestHost(
+            testCases,
+            runSettings,
+            options: null,
+            testRunEventsHandler,
+            customTestHostLauncher);
+    }
+
+    /// <inheritdoc/>
+    public void RunTestsWithCustomTestHost(
+        IEnumerable<TestCase> testCases,
+        string? runSettings,
+        TestPlatformOptions? options,
+        ITestRunEventsHandler testRunEventsHandler,
+        ITestHostLauncher? customTestHostLauncher)
+    {
+        RunTestsWithCustomTestHost(
+            testCases,
+            runSettings,
+            options,
+            testSessionInfo: null,
+            testRunEventsHandler,
+            customTestHostLauncher);
+    }
+
+    /// <inheritdoc/>
+    public void RunTestsWithCustomTestHost(
+        IEnumerable<TestCase> testCases,
+        string? runSettings,
+        TestPlatformOptions? options,
+        TestSessionInfo? testSessionInfo,
+        ITestRunEventsHandler testRunEventsHandler,
+        ITestHostLauncher? customTestHostLauncher)
+    {
+        var testCaseList = testCases.ToList();
+        _testPlatformEventSource.TranslationLayerExecutionStart(
+            1,
+            0,
+            testCaseList.Count,
+            runSettings ?? string.Empty);
+
+        try
+        {
+            TestRequestManager?.ResetOptions();
+
+            var shouldLaunchTesthost = true;
+
+            // We must avoid re-launching the test host if the test run payload already
+            // contains test session info. Test session info being present is an indicative
+            // of an already running test host spawned by a start test session call.
+            var customLauncher =
+                shouldLaunchTesthost && testSessionInfo == null
```

Comment:
smells of race condition but I can be wrong..is it resilient re-run just in case? Here and in the other places

Reply:
Oh, no, it's got nothing to do with race conditions. Basically this is just to guard against wrong usage of our wrapper. What I meant by that comment is that if we already started a test session by using the `StartTestSession` call (which spawns all the subordinate testhosts) we shouldn't re-launch testhosts if the user provides a custom testhost launcher since that won't make sense.

Reply:
ok so it's always good start more than one if without test session info